feat: define redundant rbs select task interface

wvmcastro opened the pull request.

wvmcastro committed:
feat: define redundant rbs select task interface

wvmcastro commented:
FYI @doudou

doudou commented:
> FYI @doudou 

Interface looks good. Don't forget to add a timeout to the port_driven call or the component won't be called if both inputs are dead

wvmcastro committed:
fix: port driven timeout

wvmcastro committed:
chore: redundant rbs selector state machine

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,160 @@
+    RedundantRBSSelectorTask::InternalState internal)
+{
+    // for readability's sake
+    using States = RedundantRBSSelectorTask::States;
```
You actually do not need to have the `States::` prefix ... for readability's sake ...

wvmcastro replied:
I really prefer to maintain the `States::` scope prefix

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,160 @@
+            if (now > internal.secondary_source_deadline) {
+                return updateState(States::REDUNDANCY_OFF, internal);
+            }
+        case States::HYSTERESIS_WAIT:
```
I would call that state 'MAIN_SOURCE_RECOVERING'

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,160 @@
+            }
+
+            if (now > internal.secondary_source_deadline) {
+                return updateState(States::REDUNDANCY_OFF, internal);
```
INVALID_SECONDARY_SOURCE ?

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+
+            if (now > internal.secondary_source_deadline) {
+                return updateState(States::REDUNDANCY_OFF, internal);
+            }
```
Not sure it's intended, but this goes through to the next `case`. Intended or not, fix it. Check for all conditions in all states.

wvmcastro replied:
It was my intention since the next state have its own checks

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+                return updateState(States::REDUNDANCY_OFF, internal);
+            }
+        case States::HYSTERESIS_WAIT:
+            if (now > internal.hysteresis_deadline ||
+                now > internal.secondary_source_deadline) {
+                return updateState(States::RUNNING, internal);
+            }
```
The point of this state is to verify that the main source remains valid for a long time before switching back. You need to test this

```
if (now > internal.main_source_deadline) {
    return updateState(States::INVALID_MAIN_SOURCE, internal);
}
// then the check that you've written
```

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+            if (now > internal.secondary_source_deadline) {
+                return States::NO_VALID_SOURCES;
+            }
+            if (now > internal.main_source_deadline &&
```
now < internal.main_source_deadline ?

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+                return States::NO_VALID_SOURCES;
+            }
+            if (now > internal.main_source_deadline &&
+                internal.hysteresis_deadline.isNull()) {
```
I don't get that one. If we're in INVALID_MAIN_SOURCE and the main source is back to being valid, we need to go into the hysteresis state unconditionally, no ?

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+}
+
+RedundantRBSSelectorTask::States updateState(
+    RedundantRBSSelectorTask::States nominal_state,
```
`nominal_state` ? Do you mean `current_state` ?

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+            }
+        case States::NO_VALID_SOURCES:
+        default:
+            return nominal_state;
```
This really should be a throw.

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,164 @@
+        case States::NO_VALID_SOURCES:
+            break;
+        default:
+            break;
```
`throw`. Should not happen. Make the list of states exhaustive.

wvmcastro committed:
chore: rename state HYSTERESIS_WAIT to MAIN_SOURCE_RECOVERING

wvmcastro committed:
fix: INVALID_MAIN_SOURCE transition

wvmcastro committed:
chore: throw if invalid state machine state

wvmcastro committed:
fix: compute all RUNNING state transitions

wvmcastro committed:
fix: compute all MAIN_SOURCE_RECOVERING transitions

wvmcastro committed:
chore: rename state REDUNDANCY_OFF to INVALID_SECONDARY_SOURCE

wvmcastro committed:
test: RedundantRBSSelectorTask state transitions

wvmcastro commented on test/redundant_rbs_selector_task_test.rb:
```diff
@@ -0,0 +1,172 @@
+            transition_to_main_source_recovering
+
+            expect_execution.to do
+                emit task.no_valid_sources_event
+                custom_runtime_states(except: ["no_valid_sources"]).each do |state|
+                    not_emit task.send("#{state}_event")
+                end
```
I tried writing a helper like
```suggestion
                emit_only "no_valid_source"
```
but struggled with context and/or expectations parsing (I am not sure realy). Help would be appreciated

doudou replied:
You would need to expand an intenral context from Roby/Syskit ... lots of complications for little value.

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,188 @@
+            if (now > internal.main_source_deadline) {
+                return States::NO_VALID_SOURCES;
+            }
+            if (now > internal.secondary_source_deadline) {
```
`now < `

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,188 @@
+
+bool rbsIsValid(base::samples::RigidBodyState const& rbs)
+{
+    if (rbs.hasValidPosition() && rbs.hasValidPositionCovariance() &&
+        rbs.hasValidVelocity() && rbs.hasValidVelocityCovariance() &&
+        rbs.hasValidAngularVelocity() && rbs.hasValidAngularVelocityCovariance()) {
+        return true;
+    }
```
Ignore covariances ... they  are actually invalid in our cases.

wvmcastro committed:
chore: BOTH_SOURCES_VALID state

it is equivalent to the RUNNING state which previously
indicated that both sources were fine.

This has two advantages:
  1. is clear/accurate than RUNNING
  2. its emission can be tested against at unit tests

wvmcastro committed:
fix: do not check for covariances

wvmcastro committed:
refactor: use helpers for readability bug avoidance

more than one time I switched a > by a < when comparing
now time and deadline. Besides reading isMainSoruceValid
is a lot better than now < m_internal_state.main_souce_deadline

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,222 @@
+                return States::INVALID_MAIN_SOURCE;
+            }
+
+            if (now < internal.hysteresis_deadline) {
+                if (!isSecondarySourceValid(internal)) {
+                    return States::INVALID_SECONDARY_SOURCE;
+                }
+                return States::MAIN_SOURCE_RECOVERING;
+            }
+
+            if (!isSecondarySourceValid(internal)) {
+                return States::INVALID_SECONDARY_SOURCE;
+            }
+            return States::BOTH_SOURCES_VALID;
```
The check for INVALID_SECONDARY_SOURCE is the same on both branches.

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,222 @@
+                if (!isSecondarySourceValid(internal)) {
+                    return States::NO_VALID_SOURCES;
+                }
+                else {
```
I'd remove the else (leaving the return after the if) to be consistent.

doudou commented on tasks/RedundantRBSSelectorTask.cpp:
```diff
@@ -0,0 +1,222 @@
+
+bool rbsIsValid(base::samples::RigidBodyState const& rbs)
+{
+    if (rbs.hasValidPosition() && rbs.hasValidVelocity() &&
```
missing hasValidOrientation

wvmcastro committed:
fix: check for valid orientation

wvmcastro committed:
chore: remove useless else

wvmcastro committed:
chore: refactor INVALID_SECONDARY_SOURCE transition from MAIN_SOURCE_RECOVERING
